### ISSUE TITLE
Python AutoEncoder for Color Images

### ISSUE THREAD

**Issue by schrum2** [schrum2/MM-NEAT#704]
I'd like to eventually use the MAP-Elites autoencoder on color images #703, but we have to fix the autoencoder first. Although there is a way to tweak the existing autoencoder to make it work (or at least run), it would probably be better to go all the way and implement a Convolutional Autoencoder.

Make a new subdirectory in `src\main\python` to store any work associated with this. Try to implement a Convolutional AutoEncoder in PyTorch. You could test it out by applying it to the CIFAR dataset.

Find examples online to follow. Here is one:
https://www.kaggle.com/ljlbarbosa/convolution-autoencoder-pytorch
and another:
https://analyticsindiamag.com/how-to-implement-convolutional-autoencoder-in-pytorch-with-cuda/

**Comment by annazwicker** [comment]
Using the autoencoder from the second link.  The images produced by the autoencoder somewhat matched the pictures shown by the website, neither are great.  The images below are from a run with epochs = 200 (according to the suggestion presented by the website).  Figure 1 and 2 show the original images, and Figure 3 shows the results produced by the convolutional autoencoder.
![Figure_1](https://user-images.githubusercontent.com/82399865/122825736-f105e980-d2a7-11eb-9310-2d29449ae747.png)
![Figure_2](https://user-images.githubusercontent.com/82399865/122825737-f105e980-d2a7-11eb-8c7f-2135973ec2f4.png)
![Figure_3](https://user-images.githubusercontent.com/82399865/122825738-f105e980-d2a7-11eb-8531-60d73cedbe3e.png)

**Comment by schrum2** [comment]
Since this seems to work (sort of), create a directory of color images from Picbreeder (do not commit to repo ... this is just for your local machine) and try adjusting the code to be able to train on the images in that directory. The path to the directory with the images should eventually be a command line parameter to the Python script you are writing. So far, this issue does not involve ANY Java code.

**Comment by schrum2** [comment]
By the way, this directory should consist of 28 x 28 pixel images. It is possible that this will cause problems with the existing code ... but maybe not. Convolutional NNs can be flexible with respect to input size, but this is an issue to be aware of. Just try making the directory of images and seeing what happens (having a training set of much smaller images might even greatly reduce our training time).

**Comment by annazwicker** [comment]
Differentiating between a dataset and a dataloader.  Both need to be replaced in order to load the Picbreeder training data.
https://pytorch.org/tutorials/beginner/basics/data_tutorial.html
Might need to write a custom dataset class
https://discuss.pytorch.org/t/how-to-load-images-without-using-imagefolder/59999

**Comment by annazwicker** [comment]
The following results are from the `ColorConvolutionalAutoencoderWithCUDA `with the number of epochs set to different amounts.  These runs are using the ColorTrainingSet for training the autoencoder.
Original images:
![Original](https://user-images.githubusercontent.com/82399865/123159423-9d73d700-d432-11eb-9c4d-cefb2bab2db6.PNG)
![Original2](https://user-images.githubusercontent.com/82399865/123159426-9d73d700-d432-11eb-8104-ae0423ad94ec.PNG)
200 epochs:
![200Epochs](https://user-images.githubusercontent.com/82399865/123159463-a95f9900-d432-11eb-8553-766009948cb3.PNG)
1000 epochs:
![1000Epochs](https://user-images.githubusercontent.com/82399865/123159487-b1b7d400-d432-11eb-835a-83e3e22ecce3.PNG)
5000 epochs:
![5000Epochs](https://user-images.githubusercontent.com/82399865/123159497-b54b5b00-d432-11eb-8250-1fb85ce7c0d0.PNG)
10000 epochs:
![10000Epochs](https://user-images.githubusercontent.com/82399865/123159512-b8dee200-d432-11eb-971c-2392552750b4.PNG)
50000 epochs:
![50000Epochs](https://user-images.githubusercontent.com/82399865/123159633-dca22800-d432-11eb-9117-41318e7064be.PNG)

**Comment by schrum2** [comment]
It should be possible to make this new Python code work using the exact same Java code. Specifically, all that would have to change in the Java code is the name of the python scripts being used. This means you do not need to make any new Java classes, but the variables that are currently `final` need to instead be determined by whether or not the black-and-white or color autoencoder is being used.

Basically, there can be a single boolean command line parameter "convolutionalAutoencoder" (default to false) that controls which Python scripts are used.

This fact should inform how you complete and test the python scripts.

**Comment by annazwicker** [comment]
The `colorConvolutionalAutoencoderWithCUDA `needs to be retrained check the outputs match between what is produced by Python and Java.

**Comment by schrum2** [comment]
Also, for the other python script, see what happens if you show the image (from within Python) of the input before it goes through the network. I suspect we are sending numbers in the wrong order from Java to Python.

**Comment by annazwicker** [comment]
The input image for the `ColorConvolutionalAutoencoderWithCUDA` is now correct.  TrainAutoEncoderProcess is now set up to use the both `MyAutoencoder` and `ColorConvolutionalAutoencoderWithCUDA`.

**Comment by schrum2** [comment]
If we're really at that point, then you should be able to make a batch file for a color image using the Gaier scheme with the convolutional autocoder set to true. Also, even though the skull is black and white, I would like to see a batch file that uses the convolutional autoencoder (I think in that case you would set black/white to false even though you are trying to match the skull)

**Comment by schrum2** [comment]
Running this on the skull is a priority issue

**Comment by annazwicker** [comment]
![7-6-2021ConvolutionalAutoencoderError](https://user-images.githubusercontent.com/82399865/124671605-58609380-de7b-11eb-9134-30cbc35d8676.JPG)
Error given when running the batch file SkullConvolutionalAutoencoderGaierRegularGenotype.

**Comment by schrum2** [comment]
This code runs now. I'm testing the batch file.

**Comment by schrum2** [comment]
Convolutional Autoencoder seems to recreate the Rainbow reasonably well.

Target:
![PicbreederRainbow](https://user-images.githubusercontent.com/8397236/124829911-d46fdf80-df3e-11eb-8062-25256769c7e7.jpg)

Some output from PictureTarget with the Convolutional Autoencoder:
![0 0percent0 7800516440635321fitness-Neurons24loss12](https://user-images.githubusercontent.com/8397236/124829952-e2bdfb80-df3e-11eb-9638-fdaa7e22fba3.jpg)
![0 0percent0 7842729055323144fitness-Neurons16loss8](https://user-images.githubusercontent.com/8397236/124829954-e2bdfb80-df3e-11eb-86c5-36430e8cb4fb.jpg)
![0 0percent0 7814948968145831fitness-Neurons25loss9](https://user-images.githubusercontent.com/8397236/124829955-e2bdfb80-df3e-11eb-92d2-ffdc78868e74.jpg)
![0 0percent0 7820970161933762fitness-Neurons17loss14](https://user-images.githubusercontent.com/8397236/124829956-e3569200-df3e-11eb-8170-cfcba9a2bd47.jpg)
![0 0percent0 7828919685232315fitness-Neurons24loss14](https://user-images.githubusercontent.com/8397236/124829957-e3569200-df3e-11eb-858b-52d794813295.jpg)

Still want to check with the skull though

**Comment by schrum2** [comment]
The convolutional autoencoder runs on the skull. It is clearly trying to produce black and white images with symmetry and roundness. Nothing spectacular.
![0 002551020408163265percent0 7503321551391016fitness-Neurons44loss1](https://user-images.githubusercontent.com/8397236/124988598-d2bc1f80-e003-11eb-93b3-d7f8df8461d3.jpg)
![0 0012755102040816326percent0 7536397334034015fitness-Neurons44loss0](https://user-images.githubusercontent.com/8397236/124988601-d354b600-e003-11eb-9503-750799d3a6be.jpg)
![0 00510204081632653percent0 7543312317947786fitness-Neurons48loss0](https://user-images.githubusercontent.com/8397236/124988602-d354b600-e003-11eb-9123-20fe7fae4a91.jpg)
![0 002551020408163265percent0 7538662883655343fitness-Neurons43loss0](https://user-images.githubusercontent.com/8397236/124988603-d354b600-e003-11eb-84bf-4df2d55485e2.jpg)
![0 003826530612244898percent0 7538860759477211fitness-Neurons47loss0](https://user-images.githubusercontent.com/8397236/124988604-d354b600-e003-11eb-9bc8-2078feb32749.jpg)
Would probably get slightly better results if it ran longer, but not worth waiting for. Will close issue now
